src/features/implementationProvider.ts

```diff
@@ -7,18 +7,40 @@ import AbstractSupport from './abstractProvider';
 import { FindImplementationsRequest } from '../omnisharp/protocol';
 import * as serverUtils from '../omnisharp/utils';
 import { createRequest, toLocation } from '../omnisharp/typeConversion';
-import { TextDocument, Position, CancellationToken, ImplementationProvider, ProviderResult, Definition } from 'vscode';
+import { TextDocument, Position, CancellationToken, ImplementationProvider, Definition } from 'vscode';
 
 export default class CSharpImplementationProvider extends AbstractSupport implements ImplementationProvider {
-    public provideImplementation(document: TextDocument, position: Position, token: CancellationToken): ProviderResult<Definition> {
+    public async provideImplementation(document: TextDocument, position: Position, token: CancellationToken): Promise<Definition> {
         const request = <FindImplementationsRequest>createRequest(document, position);
 
-        return serverUtils.findImplementations(this._server, request, token).then(response => {
+        const implementations = await serverUtils.findImplementations(this._server, request, token).then(response => {
             if (!response || !response.QuickFixes) {
                 return;
             }
 
             return response.QuickFixes.map(fix => toLocation(fix));
         }).catch();
+
+        // Allow language middlewares to re-map its edits if necessary.
+        try {
+            const languageMiddlewares = this._languageMiddlewareFeature.getLanguageMiddlewares();
+            let locations = implementations;
+            for (const middleware of languageMiddlewares) {
+                if (!middleware.remapLocations) {
+                    continue;
+                }
+
+                const result = await middleware.remapLocations(locations, token);
+                if (result) {
+                    locations = result;
+                }
+            }
+
+            return locations;
+        }
+        catch (error) {
+            // Something happened while remapping locations. Return the original set of locations.
+            return implementations;
+        }
```

Comment:
We could generalize this to avoid duplicating each time it is needed. I worked up something in TS Playground that we can refine - [Link](https://www.typescriptlang.org/play/?ssl=15&ssc=1&pln=16&pc=1#code/JYOwLgpgTgZghgYwgAgMJxEgNluZgD2IAKgQNYQjIDeyAvgLABQoksiKA6gVGQM4AHDgFEAJsDA16zVtHhJkAGQII8hKrUYtwcjsgAKUAgDdgo6ACUIfAK5YwAHmIA+ZBAAekEKL4GjAW2A+CCdXTWYZHXYFRQwAcxs4OIgAWTNRLAgAdzgoFA8vH2QrfzgBAEEBYBpmZDrkXBAEpIgALmQ+MChQOIBuZi1ItnkUErLK6upa+rzSgW5eQRFxMAB+AAosnn4hJDEJdoWd5YkAGmQwckp29EwIHDUiUgoQAEp2wxMzS2s7RyOlnsVs5+kx6shZmVlKp8EQ+BssCpHiA+O1ociANoAXXOlxeNww2FwsJIVzeHyMpnMUCstnsDnRJOxIIGESYYAAngIUABxCBgMYCYhckLENyeShFdbrIRQOD+fnQdoYDm4skEu4PEnPSivZAAXlcKr1AB9kDZvBAYKAIKJnNM6vrkGKCpLfDLcvLFVB2qAYNBimr8WhCfdieodW8Da5PlSfnTHH6AxZXKtishlSAOaDOdzihA5vpPQq2MLuQ4UuLCr4KByCDB83MJq4nXyBQWymWQoKJhiUliWUxmAhcHxfLEmolkmlRBlsrkIAAxCB4Gx5Gpg+oCbrGPAoPJwURELAc5AAfUCs8yOTyqKU8SnqXS14X2INyGxoIdyAENgARlgwAIMgyRgBOzTTs+863us7z3pOLQznON4QG+UybuCdR5GAa5UGAAAWQQAHQXlBKF8KC4KDBhP7-oBwFwHwHKYBCHYCBWVZusgtb1o24xVOc+icd4viCkWcoltAXYVs4zjrN+4KQkKIrtCk5w8MAcSgHAWAfEG1whpq4ZPGScGfIEwQOPoYQKfUXSnuhmFOcgCBwpIjQQU+V7QdY74EcRoHgY+SEvjBryUc54K7lArFzNyojvhpWkgDpX40ZFMA8Mg6yuSikiXshC7ILxHnBWRC58HqjmRc5uWdMgJb4QQCVOgVoWoUpXZYhFNWYcADbrAAhI1zVVbZvXgrl+AgDYEA9RNdTUQt4IAPQrWgBD+IEkhwBcIoud0CrIJlMUESgcQHmwIEEM1401XVki3n875Wc4cA5BIDX8k1ojrEp8X6W880Lf12VPfYY3pctWFsfF77g2AwMTVoC1LRN2G4bFZTxUji13TCCD4dl0BGFAkMLRjUBUEl2lYLj0g0Vo1EPeepWIeVeTLqu65OiA2TwZ5IU+VzOF5LBoLRQ0SIkr4TpnmzkHeShItrhARFKesABESmMuofCa+cDLS+ozLYobtxEsikbONQdDhcwku2p9csK15hWcyuotqxr2tsQCuwQPsYAG8gDj+ycYA23Q5uhlqEZklH9tMEAA)

Reply:
yeah it would be good to not copy paste into multiple places 😀

Reply:
Updated my type crime to actually check the parameter type - [Link](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgMJxEgNluZgD2IAKgQNYQjIDeyAvgLABQoksiKA6gVGQM4AHDgFEAJsDA16zVtHhJkAGQII8hKrUYtwcjsgAKUAgDdgo6ACUIfAK5YwAHmIA+ZBAAekEKL4GjAW2A+CCdXTWYZHXYFRQwAcxs4OIgAWTNRLAgAdzgoFA8vH2QrfzgBAEEBYBpmZDrkXBAEpIgALmQ+MChQOIBuZi1ItnkUErLK6upa+rzSgW5eQRFxMAB+AAosnn4hJDEJdoWd5YkAGmQwckp29EwIHDUiUgoQAEp2wxMzS2s7RyOlnsVs5+kx6shZmVlKp8EQ+BssCpHiA+O1ociANoAXXOlxeNww2FwsJIVzeHyMpnMUCstnsDnRJOxIIGESYYAAngIUABxCBgMYCYhckLENyeShFdZCKBwfz86DtDAc3Fkgl3B4k56UV7IAC8rmVzmmdT1yDFBUlvmluTlCqg7VAMGgxVV+LQhPuxPU2re+tcnypPzpjidLosrlWxWQSpAHNBnO5xQgc30tvlbGF3IcKXFhV8FA5BBgybmE1cZr5ApTZSzIQAckR63ZcAAjTIOQUTDEpLHOFlMZgIXB8XyxJqJZJpUQZbK5CAAMQgeBseRqYPqAm6xjwKDycFERCwHOQAH1AjPMjk8qilPFJ6l0lf59j9chsaCTcgBDZ28AEMgyRgOOzRTk+c43us7x3hOLTTrO14QK+UwbuCdR5GAq5UGAAAWQQAHTnuBiF8KC4KDKh36-lg-7IHAfAcpgEI1gIOZ5layCFsWpbjFU5z6Ox3i+IKaayhm0B1jm-brF+4KQkKIrtCk5w8MAcSgHAWAfG61wehq3pPGS0GfIEwQOPoYSyfUXQnihaH2cgCBwpIjSgY+l4QdYb64QRQEgQ+8HPpBrxkQ54I7lAzFzNyohvqp6kgJpn6UWFMA8Mg6xOSikgXgh87INxrkBcR858LqdlhQ5AD0VVoAQ-iBJIcAXCKjndPKyCthAqg2MEyBZCgZAgAQWSdfybAXDhGCTSgTn+FUmRQFZlWOc5yAZjhBCxWaDjStpjmtLcRLIr6rwGoG3w0r89IWc4uVBUh8l1lioUreCwAlusACEG1beVy1vXUWX4CANgQK9gPSIOKVvVlnTMSGb5wDkEjrfym2iOs8kxTpbwQ4DH0ZTefz-TDkNRWUMVvsT9j429WiAxRgMYVhFMCDFdN1AzDkwggOEZdARhQKTzP8qz8UaVgnNQ+RrLQ3DkinkVcElXkS4rmuZogNkMFuYFnnq5heRQaCEUNEiJK+GaSv3irHmIYbq4QPh8nrAARPJjLqHwbvnA4AK7BA+xgM41B0H7R1eidZKh3QIXMGbEArG+NuwWB9vzo7eQuyx7vyQHJxgL7yD+9sgJB8CYcR56mo+jHYfx0wQA)

Reply:
👍. I'll update the PR.